pallets/axelar-router/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+
+			Self::deposit_event(Event::<T>::InboundChainIdSet {
+				chain_id_hash,
+				chain_id,
+			});
+
+			Ok(())
+		}
+
+		#[pallet::weight(Pallet::<T>::weight_for_set_method())]
+		#[pallet::call_index(2)]
+		pub fn set_outbound_config(
+			origin: OriginFor<T>,
+			chain_id: EVMChainId,
+			config: AxelarConfig,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				T::EvmAccountCodeChecker::check((
+					config.evm.target_contract_address,
+					config.evm.target_contract_hash,
+				)),
+				Error::<T>::ContractCodeNotMatch
+			);
+
+			OutboundConfig::<T>::insert(chain_id, config.clone());
+
+			Self::deposit_event(Event::<T>::OutboundConfigSet { config: config });
+
+			Ok(())
+		}
+	}
+
+	#[precompile_utils::precompile]
+	impl<T: Config> Pallet<T> {
+		// Mimics:
+		//
+		//   function execute(
+		//         bytes32 commandId,
+		//         string calldata sourceChain,
+		//         string calldata sourceAddress,
+		//         bytes calldata payload
+		//     ) external { bytes32 payloadHash = keccak256(payload);
+		// 		 if (
+		//           !gateway.validateContractCall(
+		//              commandId,
+		//              sourceChain,
+		//              sourceAddress,
+		//              payloadHash)
+		//           ) revert NotApprovedByGateway();
+		//
+		//        _execute(sourceChain, sourceAddress, payload);
+		// }
+		//
+		// Note: The _execute logic in this case will forward all calls to the
+		//       liquidity-pools-gateway with a special runtime local origin
+		#[precompile::public("execute(bytes32,string,string,bytes)")]
+		fn execute(
+			handle: &mut impl PrecompileHandle,
+			_command_id: H256,
+			source_chain: BoundedString<ConstU32<MAX_SOURCE_CHAIN_BYTES>>,
+			source_address: BoundedString<ConstU32<MAX_SOURCE_ADDRESS_BYTES>>,
+			payload: BoundedBytes<ConstU32<MAX_PAYLOAD_BYTES>>,
+		) -> EvmResult {
+			ensure!(
+				handle.context().caller == GatewayContract::<T>::get(),
+				PrecompileFailure::Error {
+					exit_status: ExitError::Other("gateway contract address mismatch".into()),
+				}
+			);
+
+			let source_chain_id = SourceChain::<T>::get(BlakeTwo256::hash(source_chain.as_bytes()))
+				.ok_or(PrecompileFailure::Error {
+					exit_status: ExitError::Other("converter for source not found".into()),
+				})?;
+
+			let source_address_bytes =
+				cfg_utils::decode_var_source::<EXPECTED_SOURCE_ADDRESS_SIZE>(
+					source_address.as_bytes(),
+				)
+				.ok_or(PrecompileFailure::Error {
+					exit_status: ExitError::Other("invalid source address".into()),
+				})?;
+
+			let origin = DomainAddress::EVM(source_chain_id, source_address_bytes);
+
+			let msg = BoundedVec::<u8, <T::Gateway as MessageReceiver>::MaxEncodedLen>::try_from(
+				payload.as_bytes().to_vec(),
+			)
+			.map_err(|_| PrecompileFailure::Error {
+				exit_status: ExitError::Other("payload conversion".into()),
+			})?;
+
+			T::Gateway::receive(origin, msg).map_err(|e| TryDispatchError::Substrate(e).into())
```

Comment:
Probably here we should add a `router_id` parameter or append to the origin

Reply:
Yup, atm I have a `T::Hash` for inbound routers but we can change to whatever.